openedx/core/djangoapps/course_date_signals/handlers.py

```diff
@@ -118,6 +118,63 @@ def _get_custom_pacing_children(subsection, num_weeks):
     return section_date_items
 
 
+def extract_dates_from_course_spaced_out_sections(course):
+    """
+    Extract all dates from the supplied course. Apply PLS to subsections that
+    don't have custom relative_weeks_due set, by spacing them out evenly based
+    on the estimated course duration.
+    """
+    date_items = []
+    # Apply the same relative due date to all content inside a section,
+    # unless that item already has a relative date set
+    for _, section, weeks_to_complete in spaced_out_sections(course):
+        section_date_items = []
+        # section_due_date will end up being the max of all due dates of its subsections
+        section_due_date = timedelta(weeks=1)
+        for subsection in section.get_children():
+            # If custom pacing is set on a subsection, apply the set relative
+            # date to all the content inside the subsection. Otherwise
+            # apply the default Personalized Learner Schedules (PLS)
+            # logic for self paced courses.
+            relative_weeks_due = subsection.fields['relative_weeks_due'].read_from(subsection)
+            if (CUSTOM_RELATIVE_DATES.is_enabled(course.id) and relative_weeks_due):
+                section_due_date = max(section_due_date, timedelta(weeks=relative_weeks_due))
+                section_date_items.extend(_get_custom_pacing_children(subsection, relative_weeks_due))
+            else:
+                section_due_date = max(section_due_date, weeks_to_complete)
+                section_date_items.extend(_gather_graded_items(subsection, weeks_to_complete))
+        if section_date_items and (section.graded or CUSTOM_RELATIVE_DATES.is_enabled(course.id)):
+            date_items.append((section.location, {'due': section_due_date}))
+        date_items.extend(section_date_items)
+    return date_items
+
+
+def extract_dates_from_course_custom_dates_only(course):
+    """
+    Extract all dates from the supplied course. Only considers subsections that
+    have relative_weeks_due set, either custom or through Advanced Settings.
+    """
+    date_items = []
+    # Apply relative due date only to content inside a section,
+    # that already has a relative date set. Also inherits relative
+    # due date set in the advanced settings.
+    for section in course.get_children():
+        if section.visible_to_staff_only:
+            continue
+        section_date_items = []
+        for subsection in section.get_children():
+            # If custom pacing is set on a subsection, apply the set relative
+            # date to all the content inside the subsection.
+            relative_weeks_due = subsection.fields['relative_weeks_due'].read_from(subsection)
+            if relative_weeks_due:
+                section_due_date = timedelta(weeks=relative_weeks_due)
+                section_date_items.extend(_get_custom_pacing_children(subsection, relative_weeks_due))
+        if section_date_items:
```

Comment:
Don't we need to check `(section.graded or CUSTOM_RELATIVE_DATES.is_enabled(course.id))` anymore? If we do, wouldn't it be easier to move the following section to `extract_dates_from_course_spaced_out_sections`, to avoid code repetition?
```python
            if relative_weeks_due:
                section_due_date = timedelta(weeks=relative_weeks_due)
                section_date_items.extend(_get_custom_pacing_children(subsection, relative_weeks_due))
```

If we want to make this more readable, we could extract only the logic from the `for subsection in section.get_children():` loop into a separate function.

Reply:
We don't need to check for these. This code is only reachable if `CUSTOM_RELATIVE_DATES.is_enabled(course.id) == True`. And we don't care if it's graded or not, because if a subsection has custom relative due date set we always set it, and this is the function that only deals with custom due dates. You can see that it's the same in the old implementation, it just mixes custom dates and PLS.

Reply:
At the same time, we can check for these, it won't change the outcome. So your refactoring suggestion can be done. Something like this:
```python

```

I just personally don't mind code repetition in code like this, because it makes it easier to reason about this code and you are free to change the stuff in one implementation, because you know it won't blow up in another place. And then if you need to change some code (e.g. `spaced_out_sections`) in such a way that it affect only one use case, you will either need to add flags, which makes the code harder to reason about, or you have to untangle all the complicated logic to separate it :)

Reply:
At the same time, we can check for these, it won't change the outcome. So your refactoring suggestion can be done.

I just personally don't mind code repetition in code like this, because it makes it easier to reason about this code and you are free to change the stuff in one implementation, because you know it won't blow up in another place. And then if you need to change some code (e.g. `spaced_out_sections`) in such a way that it affect only one use case, you will either need to add flags, which makes the code harder to reason about, or you have to untangle all the complicated logic to separate it :)